<!DOCTYPE html>
<html lang="en-us" dir="ltr" itemscope itemtype="http://schema.org/Article">

  <head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <title>Blogue do Caloni - Antidebugging using exceptions (part one)</title>
  <meta name="author" content="" />
  <meta name="description"
        content="A debugger puts breakpoints to stop for a moment the debuggee execution. In order to do this it makes use of a well known instruction: int 3. This instruction throws an exception - the breakpoint..."/>

  <meta name="generator" content="Hugo 0.67.1" />

  
  <meta itemprop="name" content="Antidebugging using exceptions (part one)"/>
  <meta itemprop="description"
        content="A debugger puts breakpoints to stop for a moment the debuggee execution. In order to do this it makes use of a well known instruction: int 3. This instruction throws an exception - the breakpoint..."/>
  <meta itemprop="image"
        content=""/>

  
  <meta property="og:title" content="Antidebugging using exceptions (part one)"/>
  <meta property="og:type"
        content="article"/>
  <meta property="og:url" content="http://www.caloni.com.br/antidebugging-using-exceptions-part-one/"/>
  <meta property="og:image"
        content=""/>
  <meta property="og:description"
        content="A debugger puts breakpoints to stop for a moment the debuggee execution. In order to do this it makes use of a well known instruction: int 3. This instruction throws an exception - the breakpoint..."/>
  <meta property="og:site_name" content="Blogue do Caloni"/>
  <meta property="article:published_time"
        content="2008-07-28T00:00:00&#43;00:00"/>
  <meta property="article:section" content="post"/>

  
  
  <meta name="twitter:card" content="summary"/>
  <meta name="twitter:site"
        content=""/>
  <meta name="twitter:title" content="Antidebugging using exceptions (part one)"/>
  <meta name="twitter:description"
        content="A debugger puts breakpoints to stop for a moment the debuggee execution. In order to do this it makes use of a well known instruction: int 3. This instruction throws an exception - the breakpoint..."/>
  <meta name="twitter:creator"
        content=""/>
  <meta name="twitter:image:src"
        content="
                     /images/movies.png
                   "/>

  
  <link href="" rel="alternate" type="application/rss+xml"
        title="Blogue do Caloni"/>
  <link href="" rel="feed" type="application/rss+xml"
        title="Blogue do Caloni"/>

  
  <link rel="stylesheet" type="text/css" href="/css/capsule.min.css"/>
  
  <link rel="stylesheet" type="text/css" href="/css/custom.css"/>

  
  <script>
  
    
      <!-- Google Analytics -->
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
      ga('create', 'UA-50557403-2', 'auto');
      ga('send', 'pageview');
    
  
  </script>
  <script src="/js/jquery.min.js"></script>
  <script src="/js/list.js"></script>

  
  <link rel="icon" href="/favicon.ico"/>
  </head>


  <body style="min-height:100vh;display:flex;flex-direction:column">

    <nav class="navbar has-shadow is-white"
     role="navigation" aria-label="main navigation">
  <div class="container">

    <div class="navbar-brand">
      <a class="navbar-item" title="Go to Home" href="/">
        <img alt="Brand" src="/brand.svg">
        <div class="title is-6">&nbsp;Blogue do Caloni</div>
      </a>

      
      <label class="button navbar-burger is-white" for="navbar-burger-state">
        <span></span>                               
        <span></span>
        <span></span>
      </label>
    </div>

    
    <input type="checkbox" id="navbar-burger-state"/>

    <div class="navbar-menu">
      <div class="navbar-end" style="padding: 10px;">
      <big><b>
       <a title="posts dessa categoria" href="/categories/blog">blog</a>  <a title="posts dessa categoria" href="/categories/cinema">cinema</a>  <a title="posts dessa categoria" href="/categories/code">code</a> 
      </big></b>
      </div>
    </div>

  </div>
</nav>


      
      <div class="container">
        <p class="title">Antidebugging using exceptions (part one)</p>

        <p class="subtitle">

    <a href="https://github.com/Caloni/blog/blob/master/content/post/antidebugging-using-exceptions-part-one.md" title="source">
    2008-07-28
    </a>

</p>

        <div class="taglist">
            <i>Este ainda é um rascunho publicado prematuramente e está sujeito a mudanças substanciais.</i>

        </div>

        

        
          <div class="content">
            
              <p>A debugger puts breakpoints to stop for a moment the debuggee execution. In order to do this it makes use of a well known instruction: int 3. This instruction throws an exception - the breakpoint exception - that is caught by the operating system and bypassed to the handling code for this exception. For debuggee processes this code is inside the debugger. For free processes this code normally doesn&rsquo;t exist and the application simply crashs.</p>
<p>The main idea in this protection is to take care these exceptions during the application execution. Doing this, we can make use of this fact and, in the handling code, run the protected code. The solution here looks like a script interpreter. It consists basically of two threads: The first one read an instructions sequence and tells the second thread to run it step to step. In order to do this the second thread uses a small functions set with well defined code blocks. Here&rsquo;s the example in pseudocode:</p>
<pre><code>// the well-defined functions are functional blocks of code and have
// the same signature, allowing the creation of a pointer array to them
void WellDefinedFunction1( args );
void WellDefinedFunction2( args );
void WellDefinedFunction3( args );
//...
void WellDefinedFunctionN( args );

// this thread stays forever waiting execution commands from some
// well-defined function. the parameter that it receives is the function number
void ExecutionThread()
{
    // 2. ad aeternum
    while( true )
    {
        // 5. it runs some well-defined function by number
        ExecuteWellDefinedFunction( functionNumber );
    }
}

// the well-defined functions script is an integer array indicating 
// the number for the next function that is going to be called
int FunctionsToBeCalled[] = { 3, 4, 1, 2, 34, 66, 982, n };

int Start()
{
    // 1. we create the thread that is going to run commands
    CreateThread( ExecutionThread );

    // 3. for each script item (each function number)
    for( int i = 0; i &lt; sizeof(FunctionsToBeCalled); ++i )
    {
        // 4. tells the thread to run the function number N
        TellExecutionThreadToExecuteWellDefinedFunction( FunctionToBeCalled[i] );
    }

    // 6. end of execution.
    return 0;
} 
</code></pre>
<p>The protection isn&rsquo;t there yet. But it will as intrinsic part of the execution thread. All we need to do is to add a exception handling and to throw lots of int 3. The thrown exceptions are caught by a second function that runs the instruction before to returning:</p>
<pre><code>// filter exceptions that were thrown by the thread below
DWORD ExceptionFilterButExecuteWellDefinedFunction()
{
    // 5. run some well-defined function by number
    ExecuteWellDefinedFunction( number );

    return EXCEPTION_EXECUTE_HANDLER; // goes to except code
}

// this thread stays forever waiting execution commands from a 
// well-defined function. its &quot;parameter&quot; is the function number
void ExecutionThread()
{
    // 2. ad aeternum
    while( true )
    {
        __try
        {
            __asm int 3 // breakpoint exception

            // it stops the debugger if we have an attached debugger in
            // the process, or throws an exception if there is no one
        }
        __except( ExceptionFilterButExecuteWellDefinedFunction() )
        {
            // it does nothing. here is NOT where is the code (obvious, huh?)
        }

        Sleep( someTime ); // give some time
    }
} 
</code></pre>
<p>The execution thread algorithm is the same. Just the point where each instruction is executed depends to the exception throw system. Note that this exception has to be thrown in order to the next instruction run. This is fundamental, since this way nobody can just rip of the int 3 code to avoid the exception. If one does that, so no instruction will be executed at all.</p>
<p>In practice, if one tries to debug such a program one will have to deal with tons of exceptions until find out what&rsquo;s happening. Of course, as in every software protection, is&rsquo;s not definitive; it has as a purpose to make hard the reverse engineering understanding. That&rsquo;s not going to stop those who are really good doing that stuff.</p>
<p>The price paid for this protection stays on the source code visibility and understanding, compromised by the use of this technique. The programming is state machine based, and the functions are limited to some kind of behavior standard. So much smaller the code blocks inside the minifunctions, so much hard the code understanding will be.</p>
<p>The example bellow receives input through a command prompt and maps the first word typed to the function that must be called. The rest of the typed line is passed as arguments to the functions. The interpreter thread reads the user input and writes into a global string variable, at the same time the executor thread waits the string to be completed to starts the action. It was used the variable pool to let the code simpler, but the ideal would be some kind of synchronise, just like events, by example.</p>
<pre><code>/** @brief Sample demonstrating how to implemente antidebug in a code exception based.
@date jul-2007
@author Wanderley Caloni
*/
#include &lt;windows.h&gt;

#include &lt;iostream&gt;
#include &lt;map&gt;
#include &lt;sstream&gt;

#include &lt;string&gt;
#include &lt;stdlib.h&gt;

using namespace std;

// show available commands
bool Help(const string&amp;)
{

   cout &lt;&lt; &quot;AntiDebug Test Program\n&quot;
      &lt;&lt; &quot; Echo string to be printed\n&quot;
      &lt;&lt; &quot; System command [params]\n&quot;
      &lt;&lt; &quot; Quit\n\n&quot;;
   return true;
}

// run system/shell command
bool System(const string&amp; cmd)
{
   system(cmd.c_str());
   return true;
}

// print string to output
bool Echo(const string&amp; str)
{
   cout &lt;&lt; str &lt;&lt; endl;
   return true;
}

// quit program
bool Quit(const string&amp;)
{
   exit(0);
   return false;
}

// minifunctions array
bool (* (g_miniFuncs[]) )(const string&amp;) = { Help, System, Echo, Quit };

// &quot;minifunction -&gt; index&quot; mapping
map&lt;string, int&gt; g_miniFuncIdx;

// start minifunctions mapping
void InitializeMiniFuncIdx()
{
   g_miniFuncIdx[&quot;Help&quot;] = 0;
   g_miniFuncIdx[&quot;System&quot;] = 1;
   g_miniFuncIdx[&quot;Echo&quot;] = 2;
   g_miniFuncIdx[&quot;Quit&quot;] = 3;
}

// last line read from input
string g_currentLine;

// how much time are we going to wait for the next line?
const DWORD g_waitTime = 1000;

// run minifunctions
DWORD FilterException()
{
   DWORD ret = EXCEPTION_CONTINUE_EXECUTION;

   if( ! g_currentLine.empty() )
   {
      istringstream line(g_currentLine);
      g_currentLine.clear();

      string function;
      string params;

      line &gt;&gt; function;

      getline(line, params);

      // 5. run some well-defined function by number
      if( ! g_miniFuncs[g_miniFuncIdxfunction]  )
         ret = EXCEPTION_CONTINUE_SEARCH;
   }

   return ret;
}

DWORD WINAPI AntiDebugThread(PVOID)
{
   InitializeMiniFuncIdx(); // start minifunction mapping

   // 2. ad aeternum (or almost)
   while( true )

   {
      //FilterException();

      __try // the extern try waits for an exit command
      {
         __try // the intern try stays generating exceptions continuously
         {
            __asm int 3
         }
         // FilterException is the function who runs minifunctions
         __except( FilterException() )
         {
                // we can put some fake code here
         }
      }
      __except( EXCEPTION_EXECUTE_HANDLER )
      {
         break; // get out from ad aeternum (to the limbo?)
      }

      Sleep(g_waitTime);
   }

   return ERROR_SUCCESS;
}

/** and God said: 'int main!'
*/
int main()
{

   DWORD ret = ERROR_SUCCESS;
   DWORD tid = 0;
   HANDLE antiDebugThr;

   // 1. we create the thread that is going to run the commands
   antiDebugThr = CreateThread(NULL, 0, AntiDebugThread, NULL, 0, &amp;tid);;

   if( antiDebugThr )
   {
      // 3. for each item in the script (function numbers)
      while( cin )
      {
         cout &lt;&lt; &quot;Type something\n&quot;;

         // 4. tells the thread to run the function number N
         getline(cin, g_currentLine);

         if( WaitForSingleObject(antiDebugThr, g_waitTime * 2) != WAIT_TIMEOUT )
            break;
      }

      GetExitCodeThread(antiDebugThr, &amp;ret);
      CloseHandle(antiDebugThr), antiDebugThr = NULL;
   }

   // 6. end of execution.
   return (int) ret;
} 
</code></pre>
<p>The strength in this protection is to confound the attacker easily in the first steps (days, months&hellip;). Its weakness is the simplicity for the solution, since the attacker eventually realize what is going on. It is so easy that I will let it as an exercise for my readers.</p>
<p>In the next part we will se an alternative to make the code clearer and easy to use in the every day by a security software developer.</p>

            

          </div>

        <div class="taglist">
            
    
      [<a href="/categories/code">code</a>] 
      [<a href="/tags/draft">draft</a>] 
    
    <a href="https://twitter.com/intent/tweet?text=A%20debugger%20puts%20breakpoints%20to%20stop%20for%20a%20moment%20the%20debuggee%20execution.%20In%20order%20to%20do%20this%20it%20makes%20use%20of%20a%20well...%20http%3a%2f%2fwww.caloni.com.br%2fantidebugging-using-exceptions-part-one%2f from @caloni " style="text-decoration: underline;" title="Twitter">twitter</big></a>


        </div>

        <div class="navigation-bottom">
            <span style="float: right;"><a href="http://www.caloni.com.br/antidebugging-using-exceptions-part-two/">Antidebugging using exceptions (part two)<span class="icon"><i class="fa fa-arrow-right"></i></span></a></span>
<span style="float: left;"><a href="http://www.caloni.com.br/entrevista-com-o-caloni-no-do-zero-ao-mestre/"><span class="icon"><i class="fa fa-arrow-left"></i></span>Entrevista com o Caloni no &#39;Do ZERO ao MESTRE&#39;</a></span>

        </div>

      </div>

    <footer class="footer">
  <div class="container">
    <span style="float: right;"><i style="font-size: small;">Blogue do Caloni 2020. Copyright é invenção do capitalismo burguês opressor. Todos os direitos não serão reservados.</i></span>
  </div>
</footer>

  </body>

</html>
